src/options/attribute-sorting/utils.ts

```diff
@@ -1,68 +1,77 @@
 import { AttributeToken } from 'pug-lexer';
+import { SortAttributes } from './index';
 
 type CompareResult = -1 | 0 | 1;
-
-function compareByIndex(leftIndex: number, rightIndex: number): CompareResult {
-	if (leftIndex !== -1 && rightIndex === -1) {
-		return -1;
-	}
-	if (leftIndex === -1 && rightIndex !== -1) {
-		return 1;
-	}
-	const result = leftIndex - rightIndex;
-	if (result <= -1) {
-		return -1;
-	}
-	if (result >= 1) {
-		return 1;
-	}
-	return 0;
-}
+type CompareFunction = (a: any, b: any) => CompareResult;
 
 export function compareAttributeToken(
 	a: AttributeToken,
 	b: AttributeToken,
-	sortAttributes: string[],
-	moveToEnd: boolean = false
+	sortAttributes: SortAttributes,
+	sortAttributesBeginning: string[],
+	sortAttributesEnd: string[]
 ): CompareResult {
-	const sortPatterns: RegExp[] = sortAttributes.map((sort) => new RegExp(sort));
+	const sortPatternsBeginning: RegExp[] = sortAttributesBeginning.map((sort) => new RegExp(sort)).reverse();
+	const sortPatternsEnd: RegExp[] = sortAttributesEnd.map((sort) => new RegExp(sort));
 
 	const aName = a.name;
 	const bName = b.name;
 
-	let result: CompareResult = 0;
+	const aBeginningIndex: number = sortPatternsBeginning.findIndex((pattern) => pattern.test(aName));
+	const bBeginningIndex: number = sortPatternsBeginning.findIndex((pattern) => pattern.test(bName));
 
-	if (result === 0) {
-		let aIndex: number = moveToEnd ? Number.NEGATIVE_INFINITY : Number.POSITIVE_INFINITY;
-		let bIndex: number = moveToEnd ? Number.NEGATIVE_INFINITY : Number.POSITIVE_INFINITY;
-		let aFound = false;
-		let bFound = false;
-		for (let index = 0; index < sortPatterns.length; index++) {
-			const pattern = sortPatterns[index];
-			if (!aFound && pattern.test(aName)) {
-				aIndex = index;
-				aFound = true;
-			}
-			if (!bFound && pattern.test(bName)) {
-				bIndex = index;
-				bFound = true;
-			}
-			if (aFound && bFound) {
-				break;
-			}
-		}
+	const beginning = aBeginningIndex - bBeginningIndex;
+	if (beginning > 0) {
+		return -1;
+	}
+	if (beginning < 0) {
+		return 1;
+	}
+
+	const aEndIndex: number = sortPatternsEnd.findIndex((pattern) => pattern.test(aName));
+	const bEndIndex: number = sortPatternsEnd.findIndex((pattern) => pattern.test(bName));
+
+	const end = aEndIndex - bEndIndex;
+	if (end > 0) {
+		return 1;
+	}
+	if (end < 0) {
+		return -1;
+	}
 
-		result = compareByIndex(aIndex, bIndex);
+	if (sortAttributes === 'asc') {
+		if (aName > bName) {
+			return 1;
+		}
+		if (aName < bName) {
+			return -1;
+		}
+	} else if (sortAttributes === 'desc') {
+		if (aName > bName) {
+			return -1;
+		}
+		if (aName < bName) {
+			return 1;
+		}
 	}
 
-	return result;
+	return 0;
+}
+
+export function stableSort<T>(array: T[], compare: CompareFunction): any[] {
```

Comment:
Not quite sure, but isn't it `: T[]` as return type?

Reply:
To be honest, I'm new to TypeScript, I didn't knew how to properly type the method @lehni suggested.

Could you help me with this one?

Reply:
Ah okay, I can checkout your branch at evening and then suggest some types